src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs

```diff
@@ -810,20 +828,45 @@ public Task SendMailAsync(MailMessage message)
                 throw;
             }
 
+            // Only register on the CT if HandleCompletion hasn't started to ensure the CTR is disposed
+            bool lockTaken = false;
+            try
+            {
+                Monitor.TryEnter(tcs, ref lockTaken);
+                if (lockTaken && !tcs.Task.IsCompleted)
+                {
+                    ctr = cancellationToken.Register(s =>
+                    {
+                        ((SmtpClient)s).SendAsyncCancel();
+                    }, this);
+                }
+            }
+            finally
+            {
+                if (lockTaken) Monitor.Exit(tcs);
+            }
+
             // Return the task to represent the asynchronous operation
             return tcs.Task;
         }
 
-        private void HandleCompletion(TaskCompletionSource<object> tcs, AsyncCompletedEventArgs e, SendCompletedEventHandler handler)
+        private void HandleCompletion(TaskCompletionSource<object> tcs, CancellationTokenRegistration ctr, AsyncCompletedEventArgs e, SendCompletedEventHandler handler)
         {
             if (e.UserState == tcs)
             {
-                try { SendCompleted -= handler; }
-                finally
+                lock (tcs)
```

Comment:
Same here, why do we need to synchronize?

Reply:
Same lock as above, see my comment